File: company-tags/vanta/task_dependency.py
"""
Vanta phone screen: Task Dependency.

═══════════════════════════════════════════════════════════════════════════
                   INTERVIEW APPROACH — READ THIS FIRST
═══════════════════════════════════════════════════════════════════════════

PROBLEMS
--------
Q1: Given `all_tasks` (each task has an id + list of dependency ids) and
    `targets` (list of task ids), return every task that any target
    transitively depends on, in a valid execution order (a dep must
    appear before anything that depends on it). Excludes the targets
    themselves. Deduped across targets.

Q2: Given the same input + a single target id, return every task that
    transitively depends ON the target (its "ancestors" up the graph).

───────────────────────────────────────────────────────────────────────────
STEP 1 — CLARIFYING QUESTIONS (ASK 2-3 BEFORE CODING)
───────────────────────────────────────────────────────────────────────────
Interviewers score you on these. Jumping straight to code signals you're
guessing. If the prompt is sparse, ask:

    1. "Should the output include the target ids themselves, or only
        their dependencies?"
        → This impl excludes targets (common Vanta framing: "extra work
          needed before completing the targets"). Confirm.

    2. "Is the graph guaranteed to be a DAG, or should I detect cycles
        like A depends on B and B depends on A?"
        → If cycles possible: raise on detection.
        → Kahn's algorithm catches them for free
          (len(topo_order) != len(reachable_set) ⇒ cycle).

    3. "Will `all_tasks` contain tasks that no target depends on —
        unrelated islands in the graph?"
        → Almost always yes. Those tasks MUST NOT appear in the output.
        → This is what motivates the reachability (BFS) step.

    4. "For tasks at the same dependency level — say D and G both depend
        on A, B, C but neither depends on the other — do you care about
        their relative order in the output? Or is any valid topological
        order fine?"
        → Ask this BEFORE coding. Topo sort is not unique; independent
          nodes can appear in ANY order. If the interviewer doesn't care,
          say so out loud and move on. If they want a specific tiebreak
          (usually alphabetical), swap Kahn's queue for a min-heap
          (O((V+E) log V) instead of O(V+E)) — plan for that upfront so
          you don't have to refactor midway through.
        → Also flag: independent leaf nodes {A, B, C} can appear in any
          order among themselves for the same reason.

    5. "Can `targets` contain duplicates? (e.g., ['D', 'F', 'D'])"
        → If yes: the BFS seeding loop needs an `if tgt not in reachable`
          guard to avoid enqueueing the same target twice. Without the
          guard, correctness is unaffected (the inner visited check
          catches everything) but you waste one queue pop per duplicate.
        → If the interviewer says "assume no duplicates," the guard is
          dead code and can be removed.

    6. "Can a dep id reference a task not in `all_tasks`? Empty targets
        list? Duplicate ids in `all_tasks`?"
        → Usually no on all three, but ask — cheap to check, large
          downside if wrong. Empty targets should return []; missing
          deps should probably raise (see the KeyError guard).

───────────────────────────────────────────────────────────────────────────
STEP 2 — SOLUTION PROGRESSION (narrate this out loud)
───────────────────────────────────────────────────────────────────────────
The interviewer scores your reasoning, not just the final code. Walk them
through the progression:

  APPROACH A — recursive post-order DFS from each target (naive starting point)
    def dfs(node):
        if node in visited: return
        visited.add(node)
        for dep in by_id[node].dependencies:
            dfs(dep)
        if node not in target_set:
            result.append(node)   # post-order → topo order for free
    for tgt in targets: dfs(tgt)

    ✓ Reachability + ordering in ONE pass (post-order emission).
    ✓ Visited set is shared across all targets — no wasted re-traversal.
    ✗ Python recursion limit (~1000) blows up on deep dep chains.
    ✗ Cycle detection needs 3-color DFS (WHITE/GRAY/BLACK).
    Complexity: O(V + E).

  APPROACH B — two-pass, multi-source BFS + Kahn's  (see: dependencies_of_targets)
    Pass 1: ONE BFS starting from ALL targets simultaneously, sharing a
            single `reachable` set. Each node enqueued at most once.
    Pass 2: Kahn's topo sort restricted to `reachable`.

    ✓ Iterative — no recursion limit.
    ✓ Cycle detection free from Kahn's (len(order) != len(reachable)).
    ✓ Same shared-visited efficiency as approach A.
    Complexity: O(V + E) total.

Narration template:
    "The most natural first version is a recursive post-order DFS from each
     target with a shared visited set — one pass gets both reachability and
     topo order (post-order emission). Two concerns for production though:
     Python's recursion limit on deep chains, and cycle detection needs
     3-color DFS which is easy to get wrong. So I'll rewrite iteratively
     as two passes — multi-source BFS to find the reachable subgraph, then
     Kahn's for the topo sort. Kahn's gives cycle detection for free."

───────────────────────────────────────────────────────────────────────────
STEP 3 — MENTAL MODEL
───────────────────────────────────────────────────────────────────────────
The whole algorithm is a two-phase pipeline: SCOPE DOWN, then ORDER.

  Phase 1 (BFS reachability): "Which tasks are we even talking about?"
                              Prunes unrelated tasks from all_tasks so
                              phase 2 doesn't waste work on them.

  Phase 2 (Kahn's topo sort): "In what order should we run them?"
                              Also gives cycle detection for free.

Any graph problem shaped like "given a big graph + a starting set, do X
only within reach" tends to fit this shape (Course Schedule II, Clone
Graph, etc.). Q2 is Q1 with edges reversed — "what depends on me?" walks
the graph backwards.

COMPLEXITY
----------
Q1:  O(V + E)   (both approach A and approach B)
Q2:  O(V + E)
═══════════════════════════════════════════════════════════════════════════
"""

from collections import defaultdict, deque
from dataclasses import dataclass, field


@dataclass
class Task:
    id: str
    dependencies: list[str] = field(default_factory=list)


# ═══════════════════════════════════════════════════════════════════════════
# Q1 APPROACH B (production): multi-source BFS + Kahn's topo sort
# ═══════════════════════════════════════════════════════════════════════════

def dependencies_of_targets(all_tasks: list[Task], targets: list[str]) -> list[str]:
    """APPROACH B — iterative two-pass version. What to write for production.

    Motivation vs. the naive recursive DFS (approach A in the top docstring):
        Approach A is fine for scratch code but has two issues in real use:
          - Python's recursion limit (~1000) crashes on deep dep chains.
          - Cycle detection in a recursive DFS needs 3-color marking
            (WHITE/GRAY/BLACK), which is easy to get subtly wrong.
        Rewriting iteratively as two passes fixes both — and Kahn's gives
        us cycle detection for free (len(order) != len(reachable)).

    Mental model — SCOPE DOWN then ORDER:
        Pass 1: Multi-source BFS builds `reachable`, the induced subgraph
                of tasks the targets transitively depend on. `reachable`
                doubles as the BFS visited set — shared across all targets,
                so overlapping subgraphs are walked exactly once.
        Pass 2: Kahn's topological sort restricted to `reachable`.
        Pass 3: Drop targets from the final list (per problem framing).

    Complexity: O(V + E) total.
    """
    by_id: dict[str, Task] = {t.id: t for t in all_tasks}
    target_set: set[str] = set(targets)

    for tgt in targets:
        if tgt not in by_id:
            raise KeyError(f"target {tgt!r} not found in all_tasks")

    # ─── PASS 1: Multi-source BFS (SHARED visited set) ─────────────────
    # `reachable` does double duty: (a) the visited set that prevents
    # re-processing shared subgraphs and cycles, (b) the final set of
    # in-scope tasks that pass 2 will topo-sort.
    #
    # Seeding: enqueue every target's direct deps upfront (not the targets
    # themselves — we want to exclude targets from output). Then a single
    # BFS drains the queue. The `if dep not in reachable` guard means each
    # node enters the queue at most once, regardless of how many paths
    # reach it.
    #
    # Edge case worth mentioning: if targets = [F, D] and F depends on D,
    # then D ends up in `reachable` (F pulls it in). Its own deps then get
    # explored during BFS. We drop D from the final output in step 3.
    reachable: set[str] = set()
    queue: deque[str] = deque()
    for tgt in targets:
        for dep in by_id[tgt].dependencies:
            if dep not in reachable:
                reachable.add(dep)
                queue.append(dep)
    while queue:
        node = queue.popleft()
        for dep in by_id[node].dependencies:
            if dep not in reachable:
                reachable.add(dep)
                queue.append(dep)

    # ─── PASS 2: Kahn's topological sort restricted to `reachable` ─────
    # Kahn's gives us:
    #   (1) a valid topo order (deps before dependents), and
    #   (2) cycle detection for free.
    reverse_adj: dict[str, list[str]] = defaultdict(list)
    in_degree: dict[str, int] = {node: 0 for node in reachable}
    for node in reachable:
        for dep in by_id[node].dependencies:
            if dep in reachable:
                reverse_adj[dep].append(node)
                in_degree[node] += 1

    queue = deque(n for n, d in in_degree.items() if d == 0)
    order: list[str] = []
    while queue:
        node = queue.popleft()
        order.append(node)
        for dependent in reverse_adj[node]:
            in_degree[dependent] -= 1
            if in_degree[dependent] == 0:
                queue.append(dependent)

    if len(order) != len(reachable):
        raise ValueError("cycle detected in dependency graph")

    # ─── PASS 3: Filter targets from output ────────────────────────────
    return [n for n in order if n not in target_set]


# ═══════════════════════════════════════════════════════════════════════════
# Q1 VARIANT: same problem, but INCLUDE the targets in the output
# ═══════════════════════════════════════════════════════════════════════════

def dependencies_of_targets_include(all_tasks: list[Task], targets: list[str]) -> list[str]:
    """Variant of `dependencies_of_targets` that INCLUDES the target ids in
    the output, in valid topo order alongside their deps.

    Use this when the interviewer answers clarifying question #1 with
    "include the targets" instead of "just their deps."

    Two changes vs the exclude version:
        1. BFS init seeds with the TARGETS THEMSELVES, not their deps.
           (This ensures every target lands in `reachable`, including
            targets that no other target depends on.)
        2. No filter step at the end — return the topo order as-is.

    Complexity: O(V + E) — same as the exclude version.
    """
    by_id: dict[str, Task] = {t.id: t for t in all_tasks}
    for tgt in targets:
        if tgt not in by_id:
            raise KeyError(f"target {tgt!r} not found in all_tasks")

    # PASS 1: Multi-source BFS — seed with targets THEMSELVES.
    reachable: set[str] = set()
    queue: deque[str] = deque()
    for tgt in targets:
        if tgt not in reachable:
            reachable.add(tgt)
            queue.append(tgt)
    while queue:
        node = queue.popleft()
        for dep in by_id[node].dependencies:
            if dep not in reachable:
                reachable.add(dep)
                queue.append(dep)

    # PASS 2: Kahn's — identical to exclude version.
    reverse_adj: dict[str, list[str]] = defaultdict(list)
    in_degree: dict[str, int] = {node: 0 for node in reachable}
    for node in reachable:
        for dep in by_id[node].dependencies:
            if dep in reachable:
                reverse_adj[dep].append(node)
                in_degree[node] += 1

    queue = deque(n for n, d in in_degree.items() if d == 0)
    order: list[str] = []
    while queue:
        node = queue.popleft()
        order.append(node)
        for dependent in reverse_adj[node]:
            in_degree[dependent] -= 1
            if in_degree[dependent] == 0:
                queue.append(dependent)

    if len(order) != len(reachable):
        raise ValueError("cycle detected in dependency graph")

    # PASS 3: No filter — targets are already in the topo order in the
    # correct position (after their deps).
    return order


# ═══════════════════════════════════════════════════════════════════════════
# Q2: transitive ancestors of a single target
# ═══════════════════════════════════════════════════════════════════════════

def ancestors_of_target(all_tasks: list[Task], target: str) -> list[str]:
    """Q2 is Q1 with the edges reversed.

    Q1 asks: "what does target DEPEND on?" — walk forward along dep edges.
    Q2 asks: "what DEPENDS on target?" — walk backward along dep edges.

    Trick: build the REVERSE adjacency list once (`dep → list of dependents`),
    then BFS from the target. No topo sort needed — the problem doesn't ask
    for order among ancestors, so plain BFS discovery order is fine.

    Complexity: O(V + E) — build reverse adj is O(V + E), BFS is O(V + E).

    If the interviewer wants a specific order among ancestors, confirm
    what they want (topological? alphabetical?) before coding.
    """
    by_id: dict[str, Task] = {t.id: t for t in all_tasks}
    if target not in by_id:
        raise KeyError(f"target {target!r} not found in all_tasks")

    # ─── Build reverse adjacency: dep → list of things that depend on it ─
    # WHY: The input gives us "who I depend on" per task. To answer "who
    # depends on ME," we need to flip the edges once and then walk normally.
    reverse_adj: dict[str, list[str]] = defaultdict(list)
    for task in all_tasks:
        for dep in task.dependencies:
            reverse_adj[dep].append(task.id)

    # ─── BFS from target along the reverse edges ───────────────────────
    # `seen` starts with target because we don't want target in the output
    # (we're returning ancestors, which by definition excludes self).
    ancestors: list[str] = []
    seen: set[str] = {target}
    queue: deque[str] = deque([target])
    while queue:
        node = queue.popleft()
        for parent in reverse_adj[node]:
            if parent not in seen:
                seen.add(parent)
                ancestors.append(parent)
                queue.append(parent)

    return ancestors


# ═══════════════════════════════════════════════════════════════════════════
# demo — verify both Q1 versions produce valid topo orders
# ═══════════════════════════════════════════════════════════════════════════

def _validate_topo(output: list[str], tasks_by_id: dict[str, Task], expected: list[str]) -> str:
    """Return 'PASS' if `output` (a) contains exactly the expected set of ids
    and (b) is a valid topological order (every dep appears before its
    dependent). Otherwise a diagnostic string.

    Topo order isn't unique, so we can't string-match — we check the two
    properties that DEFINE correctness instead.
    """
    if set(output) != set(expected):
        missing = sorted(set(expected) - set(output))
        extra = sorted(set(output) - set(expected))
        return f"FAIL (content) — missing={missing}, extra={extra}"
    positions = {n: i for i, n in enumerate(output)}
    for tid in positions:
        task = tasks_by_id.get(tid)
        if task is None:
            continue
        for dep in task.dependencies:
            if dep in positions and positions[dep] >= positions[tid]:
                return f"FAIL (order) — {dep!r} must appear before {tid!r} in {output}"
    return "PASS"


if __name__ == "__main__":
    # Fixture-driven demo. Cases live in test_fixtures.json alongside this
    # file. The runner loads the fixture, executes each case against both
    # Q1 variants (exclude, include) and Q2, and validates results by
    # (a) set equality vs expected, (b) topological validity.
    #
    # Topological order is NOT unique — if A, B, C all have no deps, any
    # order among them is valid. That's why validation checks content +
    # topo validity, not string equality.
    import json
    from pathlib import Path

    fixture_path = Path(__file__).parent / "test_fixtures.json"
    with open(fixture_path) as f:
        fixtures = json.load(f)

    total, passed = 0, 0

    for graph_spec in fixtures["graphs"]:
        print("=" * 76)
        print(f"GRAPH: {graph_spec['name']}")
        print(f"  {graph_spec['description']}")
        for line in graph_spec.get("ascii_art", []):
            print(f"  {line}")
        print("=" * 76)

        tasks = [Task(t["id"], t.get("deps", [])) for t in graph_spec["tasks"]]
        tasks_by_id = {t.id: t for t in tasks}

        for case in graph_spec.get("q1_cases", []):
            targets = case["targets"]
            print(f"\nQ1 targets = {targets}")
            if case.get("notes"):
                print(f"   notes: {case['notes']}")

            for label, fn, expected in [
                ("exclude", dependencies_of_targets,         case["expected_exclude"]),
                ("include", dependencies_of_targets_include, case["expected_include"]),
            ]:
                actual = fn(tasks, targets)
                status = _validate_topo(actual, tasks_by_id, expected)
                total += 1
                if status == "PASS":
                    passed += 1
                print(f"   {label}: {actual}  [{status}]")

        for case in graph_spec.get("q2_cases", []):
            target = case["target"]
            actual = ancestors_of_target(tasks, target)
            expected = case["expected_ancestors"]
            # Q2 doesn't require topo order — the problem statement doesn't
            # specify an order among ancestors. We validate set equality only.
            if set(actual) == set(expected):
                status = "PASS"
                passed += 1
            else:
                missing = sorted(set(expected) - set(actual))
                extra = sorted(set(actual) - set(expected))
                status = f"FAIL — missing={missing}, extra={extra}"
            total += 1
            print(f"\nQ2 target = {target!r}")
            if case.get("notes"):
                print(f"   notes: {case['notes']}")
            print(f"   ancestors: {actual}  [{status}]")

    print()
    print("=" * 76)
    print(f"RESULTS: {passed}/{total} passed")
    print("=" * 76)

File: company-tags/vanta/task_dependency.cpp
// C++ mirror of task_dependency.py — same algorithms, structured the same way,
// so you can diff them mentally to learn Python idioms coming from C++.
//
// Python  ↔  C++ cheat sheet used below (with `using namespace std;` in effect):
//   list[T]                         vector<T>
//   dict[K, V]                      unordered_map<K, V>
//   set[T]                          unordered_set<T>
//   deque[T]  (used as FIFO)        queue<T>            // plain FIFO — one end only
//   defaultdict(list)               unordered_map<K, vector<V>>
//                                     (operator[] auto-inserts a default-constructed value,
//                                      same as defaultdict semantics)
//   @dataclass                      plain struct with public fields
//   x in s                          s.count(x)                    (or s.contains(x) in C++20)
//   for x in xs                     for (const T& x : xs)
//   for k, v in d.items()           iterate as pair<const K, V>; use .first / .second
//   raise KeyError("...")           throw runtime_error("...")
//   [n for n in xs if pred(n)]      range loop + push_back
//   set(xs)                         unordered_set<T>(xs.begin(), xs.end())
//   queue.append(x)                 q.push(x)
//   queue.popleft()                 t = q.front(); q.pop();       // two steps, pop() returns void
//
// Note: this file uses fully explicit types (no `auto`) to keep the C++↔Python
// mapping as legible as possible. In real C++ you'd lean on `auto` and structured
// bindings for the same loops.
//
// Complexity is identical to the Python version:
//   dependencies_of_targets_brute: O(k*(V+E)) reachability + O(V+E) topo
//   dependencies_of_targets:       O(V+E) total
//   ancestors_of_target:           O(V+E)

#include <iostream>
#include <queue>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct Task {
    string id;
    vector<string> dependencies;
};


// ---------- Q1 (brute force): per-target reachability + topo sort ----------

vector<string> dependencies_of_targets_brute(
    const vector<Task>& all_tasks,
    const vector<string>& targets
) {
    // by_id: dict[str, Task*] — pointer avoids copying the Task; the vector owns storage.
    unordered_map<string, const Task*> by_id;
    for (const Task& t : all_tasks) by_id[t.id] = &t;

    unordered_set<string> target_set(targets.begin(), targets.end());
    for (const string& tgt : targets) {
        if (by_id.find(tgt) == by_id.end()) {
            throw runtime_error("target not found in all_tasks: " + tgt);
        }
    }

    // Pass 1: one BFS *per target*, fresh visited set each time, then union.
    // Wasteful when targets share deps — same subgraph gets walked more than once.
    unordered_set<string> reachable;
    for (const string& tgt : targets) {
        unordered_set<string> per_target_seen;
        queue<string> q;
        for (const string& dep : by_id[tgt]->dependencies) {
            if (!per_target_seen.count(dep)) {
                per_target_seen.insert(dep);
                q.push(dep);
            }
        }
        while (!q.empty()) {
            string node = q.front();
            q.pop();
            for (const string& dep : by_id[node]->dependencies) {
                if (!per_target_seen.count(dep)) {
                    per_target_seen.insert(dep);
                    q.push(dep);
                }
            }
        }
        for (const string& n : per_target_seen) reachable.insert(n);
    }

    // Pass 2: Kahn's topo sort restricted to `reachable`.
    unordered_map<string, vector<string>> reverse_adj;
    unordered_map<string, int> in_degree;
    for (const string& node : reachable) in_degree[node] = 0;
    for (const string& node : reachable) {
        for (const string& dep : by_id[node]->dependencies) {
            if (reachable.count(dep)) {
                reverse_adj[dep].push_back(node);
                in_degree[node] += 1;
            }
        }
    }

    queue<string> q;
    for (const pair<const string, int>& entry : in_degree) {
        if (entry.second == 0) q.push(entry.first);
    }
    vector<string> order;
    while (!q.empty()) {
        string node = q.front();
        q.pop();
        order.push_back(node);
        for (const string& dependent : reverse_adj[node]) {
            if (--in_degree[dependent] == 0) q.push(dependent);
        }
    }

    if (order.size() != reachable.size()) {
        throw runtime_error("cycle detected in dependency graph");
    }

    vector<string> result;
    for (const string& n : order) {
        if (!target_set.count(n)) result.push_back(n);
    }
    return result;
}


// ---------- Q1 (optimized): multi-source BFS with shared visited set ----------

vector<string> dependencies_of_targets(
    const vector<Task>& all_tasks,
    const vector<string>& targets
) {
    unordered_map<string, const Task*> by_id;
    for (const Task& t : all_tasks) by_id[t.id] = &t;

    unordered_set<string> target_set(targets.begin(), targets.end());
    for (const string& tgt : targets) {
        if (by_id.find(tgt) == by_id.end()) {
            throw runtime_error("target not found in all_tasks: " + tgt);
        }
    }

    // Pass 1: multi-source BFS. `reachable` is shared across all targets, so
    // each node is enqueued at most once regardless of how many targets reach it.
    unordered_set<string> reachable;
    queue<string> q;
    for (const string& tgt : targets) {
        for (const string& dep : by_id[tgt]->dependencies) {
            if (!reachable.count(dep)) {
                reachable.insert(dep);
                q.push(dep);
            }
        }
    }
    while (!q.empty()) {
        string node = q.front();
        q.pop();
        for (const string& dep : by_id[node]->dependencies) {
            if (!reachable.count(dep)) {
                reachable.insert(dep);
                q.push(dep);
            }
        }
    }

    // Pass 2: Kahn's topo sort restricted to `reachable`. Identical to brute version.
    unordered_map<string, vector<string>> reverse_adj;
    unordered_map<string, int> in_degree;
    for (const string& node : reachable) in_degree[node] = 0;
    for (const string& node : reachable) {
        for (const string& dep : by_id[node]->dependencies) {
            if (reachable.count(dep)) {
                reverse_adj[dep].push_back(node);
                in_degree[node] += 1;
            }
        }
    }

    queue<string> topo_q;
    for (const pair<const string, int>& entry : in_degree) {
        if (entry.second == 0) topo_q.push(entry.first);
    }
    vector<string> order;
    while (!topo_q.empty()) {
        string node = topo_q.front();
        topo_q.pop();
        order.push_back(node);
        for (const string& dependent : reverse_adj[node]) {
            if (--in_degree[dependent] == 0) topo_q.push(dependent);
        }
    }

    if (order.size() != reachable.size()) {
        throw runtime_error("cycle detected in dependency graph");
    }

    vector<string> result;
    for (const string& n : order) {
        if (!target_set.count(n)) result.push_back(n);
    }
    return result;
}


// ---------- Q2: transitive ancestors of a single target ----------

vector<string> ancestors_of_target(
    const vector<Task>& all_tasks,
    const string& target
) {
    unordered_map<string, const Task*> by_id;
    for (const Task& t : all_tasks) by_id[t.id] = &t;
    if (by_id.find(target) == by_id.end()) {
        throw runtime_error("target not found in all_tasks: " + target);
    }

    // Build reverse adjacency: dep → list of things that depend on it.
    unordered_map<string, vector<string>> reverse_adj;
    for (const Task& task : all_tasks) {
        for (const string& dep : task.dependencies) {
            reverse_adj[dep].push_back(task.id);
        }
    }

    vector<string> ancestors;
    unordered_set<string> seen = {target};
    queue<string> q;
    q.push(target);
    while (!q.empty()) {
        string node = q.front();
        q.pop();
        for (const string& parent : reverse_adj[node]) {
            if (!seen.count(parent)) {
                seen.insert(parent);
                ancestors.push_back(parent);
                q.push(parent);
            }
        }
    }
    return ancestors;
}


// ---------- demo ----------

static void print_vec(const string& label, const vector<string>& v) {
    cout << label << ": [";
    for (size_t i = 0; i < v.size(); ++i) {
        cout << "'" << v[i] << "'";
        if (i + 1 < v.size()) cout << ", ";
    }
    cout << "]\n";
}

int main() {
    //
    //   A ──┐
    //   B ──┤─→ D ─→ F
    //   C ──┘       ↑
    //          E ───┘
    //
    vector<Task> tasks = {
        {"A", {}},
        {"B", {}},
        {"C", {}},
        {"D", {"A", "B", "C"}},
        {"E", {}},
        {"F", {"D", "E"}},
    };

    print_vec("Q1 brute  deps of [F]   ", dependencies_of_targets_brute(tasks, {"F"}));
    print_vec("Q1 opt    deps of [F]   ", dependencies_of_targets(tasks, {"F"}));
    print_vec("Q1 brute  deps of [D,E] ", dependencies_of_targets_brute(tasks, {"D", "E"}));
    print_vec("Q1 opt    deps of [D,E] ", dependencies_of_targets(tasks, {"D", "E"}));
    print_vec("Q1 brute  deps of [F,D] ", dependencies_of_targets_brute(tasks, {"F", "D"}));
    print_vec("Q1 opt    deps of [F,D] ", dependencies_of_targets(tasks, {"F", "D"}));
    print_vec("Q2  ancestors of A      ", ancestors_of_target(tasks, "A"));
    print_vec("Q2  ancestors of E      ", ancestors_of_target(tasks, "E"));
    print_vec("Q2  ancestors of F      ", ancestors_of_target(tasks, "F"));

    return 0;
}
